### PR TITLE
i#1312 AVX-512 support: Add test for instr_t and opnd_t blackboxes.

### PR DESCRIPTION
**Pull request by hgreving2304**
Adds a test for exported instr_t and opnd_t blackbox to match internal size.

Issue: #1312

### CODE REVIEW

File: suite/tests/client-interface/blackbox.c [hunk 1]
```diff
@@ -0,0 +1,41 @@
+/* *******************************************************************************
+ * Copyright (c) 2011 Massachusetts Institute of Technology  All rights reserved.
```
Comment (hgreving2304) [inline suite/tests/client-interface/blackbox.c]: Have to fix copyright, Google, Inc 2019.

File: suite/tests/client-interface/blackbox.c [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * DAMAGE.
+ */
+
+#include "configure.h"
+#include <stdio.h>
+
+int
+main(void)
+{
+    printf("Done\n");
```
Comment (derekbruening) [inline suite/tests/client-interface/blackbox.c]: I think the appdll should be replaced with a private-only dll.  Then this file can go away as well in favor of the shared app.

File: suite/tests/CMakeLists.txt [hunk 3]
```diff
@@ -2121,6 +2121,12 @@ if (CLIENT_INTERFACE)
       "" "")
   endif (NOT ARM)
 
+  tobuild_appdll(client.blackbox client-interface/blackbox.c)
+  configure_DynamoRIO_client(client.blackbox.appdll)
+  tobuild_ci(client.blackbox client-interface/blackbox.c "" "" "")
+  target_link_libraries(client.blackbox.dll client.blackbox.appdll)
```
Comment (derekbruening) [inline suite/tests/CMakeLists.txt]: But an appdll is an *app* dll.  It's imported by the app executable and loaded by the app loader.  Does this then result in two copies, one for the app and one private copy?  That doesn't seem ideal, and it's non-obvious.  Please add comments explaining the intent here.

I think it would be cleaner to not have this be an appdll.

File: suite/tests/client-interface/blackbox.dll.c [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Test exposed instr_t size to be identical to internal size by initializing an stack
```
Comment (derekbruening) [inline suite/tests/client-interface/blackbox.dll.c]: grammar: s/an/a/

File: suite/tests/client-interface/blackbox.dll.c [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+ * object with DynamoRIO and checking its bounds. TODO: Add test for opnd_t.
```
Comment (derekbruening) [inline suite/tests/client-interface/blackbox.dll.c]: Wait, I don't see any stack objects.  And there is a test of opnd_t.  This comment is all wrong.

File: suite/tests/client-interface/blackbox.dll.c (hunk 4 above)
Comment (hgreving2304) [reply suite/tests/client-interface/blackbox.dll.c]: This comment is stale, will fix.

File: suite/tests/client-interface/blackbox.dll.c (hunk 5 above)
Comment (hgreving2304) [reply suite/tests/client-interface/blackbox.dll.c]: Stale. Will fix.

File: suite/tests/CMakeLists.txt (hunk 3 above)
Comment (hgreving2304) [reply suite/tests/CMakeLists.txt]: No, there's only 1 copy. I am linking the app dll to the client dll. I am just using the existing function that calls it "app dll". Just to understand, by "private" dll above or by "not app dll", what do you mean? It will still be a dll linked to the client dll yes?

Comment (derekbruening) [reply suite/tests/CMakeLists.txt]: tobuild_appdll links the dll to the app executable (in most cases -- maybe it doesn't in certain cases where the app dynamically loads it, did not check the code), so I would expect two copies: one loaded by the app loader and one loaded by DR's private loader.  Hence an app dll and a private dll.

File: suite/tests/client-interface/blackbox.dll.c (hunk 4 above)
Comment (hgreving2304) [reply suite/tests/client-interface/blackbox.dll.c]: Done.

File: suite/tests/client-interface/blackbox.c (hunk 2 above)
Comment (hgreving2304) [reply suite/tests/client-interface/blackbox.c]: Done.

File: suite/tests/CMakeLists.txt (hunk 3 above)
Comment (hgreving2304) [reply suite/tests/CMakeLists.txt]: See above.

File: suite/tests/CMakeLists.txt [hunk 6]
```diff
@@ -2121,6 +2121,16 @@ if (CLIENT_INTERFACE)
       "" "")
   endif (NOT ARM)
 
+  add_library(client.blackbox.privdll SHARED client-interface/blackbox.privdll.c)
+  configure_DynamoRIO_client(client.blackbox.privdll)
+  add_library(client.blackbox.dll SHARED client-interface/blackbox.dll.c)
+  configure_DynamoRIO_client(client.blackbox.dll)
+  target_link_libraries(client.blackbox.dll client.blackbox.privdll)
+  set_target_properties(client.blackbox.dll PROPERTIES SKIP_BUILD_RPATH OFF)
```
Comment (derekbruening) [inline suite/tests/CMakeLists.txt]: Why is this necessary, and in particular for just one of the clients?  Please explain in a comment as this seems unusual.

File: suite/tests/CMakeLists.txt [hunk 7]
```diff
@@ -2121,6 +2121,16 @@ if (CLIENT_INTERFACE)
...
+  torunonly_ci(client.blackbox ${ci_shared_app} client.blackbox.dll
```
Comment (derekbruening) [inline suite/tests/CMakeLists.txt]: This is missing key items from tobuild_ci: debug link flags; dep on api_headers; copy to Android device.  Best to pull that out into a shared function and use it here.

File: suite/tests/CMakeLists.txt [hunk 8]
```diff
@@ -2121,6 +2121,16 @@ if (CLIENT_INTERFACE)
       "" "")
   endif (NOT ARM)
 
+  add_library(client.blackbox.privdll SHARED client-interface/blackbox.privdll.c)
```
Comment (derekbruening) [inline suite/tests/CMakeLists.txt]: I find the name confusing: it implies that blackbox.dll.c is *not* private.  blackbox.dll{1,2} or blackbox.dll{A,B} or sthg seems clearer.

**Comment by hgreving2304** [comment]
PTAL.

File: suite/tests/CMakeLists.txt (hunk 6 above)
Comment (hgreving2304) [reply suite/tests/CMakeLists.txt]: Done.

File: suite/tests/CMakeLists.txt (hunk 7 above)
Comment (hgreving2304) [reply suite/tests/CMakeLists.txt]: Done.
